### DISCUSSION TITLE
MixedChat_Agents - recreating conversation

### DISCUSSION THREAD

**Discussion by JP-droidic** [microsoft/semantic-kernel#8657] [Q&A] [answered]
Hi Team,

looking at https://github.com/microsoft/semantic-kernel/blob/main/dotnet/samples/Concepts/Agents/MixedChat_Agents.cs 
and wondering how you re-create conversation.

In a classic web scenario, when a user asks something, we post the question to the server, return the answer to the user, and the user asks again - we need to recreate AgentGroupChat on each post.

How should we store the chat conversation between server runs and re-create history? 

If one of the agents is an OpenAI assistant, is every (post to the server) run on a different OPENAI thread? Does the SK create a new thread and somehow fills the assistant's part of the chat history?

**Comment by crickman** [comment, marked as answer]
We are currently working on a feature to serialize and deserialize `AgentChat` (which `AgentGroupChat` will fully support).  If your scenario is bound to a mixed agent scenario, then this feature is likely required.  We hope to have this available within the next few weeks.

If you are using one agent type (only `ChatCompletionAgent` or `OpenAIAssistantAgent`), then you could preserve either the `ChatHistory` or _ThreadId_ in storage for retreival.  If your service has the concept of a user-session, then you'd retrieve the appropriate state according to that _session-key_ (which is likely component of the URI in your REST API).  

> Note: For the `ChatHistory` case, I might store each message as separate document.  We have a legacy project that managed chat history in this fashion: https://github.com/microsoft/chat-copilot

When `AgentChat` serialization feature is ready, I suspect you'd retrieve this chat payload using the same approach (by some _session-key_).  The good news is that since an single agent is able to participate in multiple `AgentChat` instances, you can include your agents in the DI pattern for injection into the `AgentChat` you'll be re-creating from the serialized payload (along with any new `AgentChat` instance).

If you must absolutely become unblocked immediately, you might consider retrieving the `AgentChat` history for storage via `GetChatMessagesAsync` and then restoring them with `AddChatMessage`...againt, using whatever _session-key_ you've defined.  For assistant, this will create a new threadid, but it will have the history present.

**Reply by JP-droidic** [reply]
Thank you @crickman, looking forward to the serialization feature
